llpc/lower/llpcSpirvLowerCfgMerges.cpp

```diff
@@ -0,0 +1,585 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2022 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  llpcSpirvLowerCfgMerges.cpp
+ * @brief LLPC source file: contains implementation of class Llpc::SpirvLowerCfgMerges.
+ * @details This pass process loop merge operations defined in SPIRV.
+ *          It manipulates loops with subgroup operations to ensure the set of lanes leaving a loop through merges,
+ *          breaks and returns are maximal reconvergent subsets.
+ ***********************************************************************************************************************
```

Comment:
We need document possible TODO improvements:
1. If we find out that all loop exit conditions are uniform, we don't need to reconverge at the loop merge.
2. The basic blocks on the loop-break path that are after the block with convergent operation can also be pulled out of the loop.

Reply:
Thanks, I have documented these.